### PR TITLE
Refactor expression node resolution logic in code actions

### LINKED ISSUE #33783: Type Cast Code action is not suggested for Record Field
**Description:**
Consider the following source code
codeac

The Type Cast for Assignment Code action is not suggested as a quick fix.
We can add the type cast manually to get the following result.

resulkt

**Steps to reproduce:**
```Ballerina
public type MyType record {|int field1;int field2;|};
public function main() {
    record {|int field1;int field2;|} a =  <MyType> {field1: 10,field2: 1.0};
}
```

**Affected Versions:**
slbeta5-snapshot

### LINKED ISSUE #34027: Add check code action does not perform a text edit
**Description:**
capture

As depicted above, the Add Check code action does not perform a text edit when it is not enclosed by a function.

**Steps to reproduce:**
```Ballerina
string[][] lines = myFunc();

function myFunc() returns string[][] | error {
    return [[]];
}
```

**Affected Versions:**
slbeta5-snapshot

### PR DESCRIPTION
**Pull request by malinthar**
## Purpose
Add ExpressionNodeResolver which is able to resolve the containing expression node of a given node if present. 

Fixes #33783 
Fixes #34027 

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [x] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Review by gayalkuruppu — COMMENTED** [review]
```
public type MyType record {|int field1;int field2;|};
public function main() {
    decimal aa = 1.0;
    record {|int field1;int field2;|} a =  <MyType> {field1: 10,field2: <float> aa};
}
```
WDYT about providing typecast code action at `<float> aa`? not seeing any use case though, but it is syntactically correct

**Comment by malinthar** [comment]
> ```
> public type MyType record {|int field1;int field2;|};
> public function main() {
>     decimal aa = 1.0;
>     record {|int field1;int field2;|} a =  <MyType> {field1: 10,field2: <float> aa};
> }
> ```
> 
> WDYT about providing typecast code action at `<float> aa`? not seeing any use case though, but it is syntactically correct

Since there is already a cast expression  at the depicted range, we refrain from suggesting the Type Cast codeaction.

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/AddCheckCodeAction.java [hunk 1]
```diff
@@ -63,8 +63,11 @@ public AddCheckCodeAction() {
         if (!DIAGNOSTIC_CODES.contains(diagnostic.diagnosticInfo().code())) {
             return Collections.emptyList();
         }
-        Node matchedNode = getMatchedNode(positionDetails.matchedNode());
-        if (matchedNode == null) {
+
+        //Check if there is a check expression already present.
+        ExpressionNodeResolver expressionResolver = new ExpressionNodeResolver();
```
Comment (nadeeshaan) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/AddCheckCodeAction.java:68]: Shall we rename the `ExpressionNodeResolver` to a bit of a specific one? This is generic

**Review by nadeeshaan — CHANGES_REQUESTED** [review]

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/AddCheckCodeAction.java (hunk 1 above)
Comment (malinthar) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/AddCheckCodeAction.java:68]: What do you recommend? Something like `MacthedExpressionNodeResolver`?

Comment (nadeeshaan) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/AddCheckCodeAction.java:68]: +1

**Review by nadeeshaan — APPROVED** [review]
LGTM

**Review by IMS94 — APPROVED** [review]

**Review by gayalkuruppu — APPROVED** [review]
LGTM

**Pull request merged**